crates/c/src/lib.rs

```diff
@@ -441,6 +453,61 @@ impl C {
             }
         }
     }
+
+    /// Removes all types from `self.{dtor_funcs,type_names,resources}` which
+    /// are redefined in exports.
+    ///
+    /// WIT interfaces can be both imported and exported but they're represented
+    /// with the same `TypeId` internally within the `wit-parser`
+    /// representation. This means that duplicate types need to be generated for
+    /// exports, even if the same interface was already imported. If nothing
+    /// were done here though then the same type imported and exported wouldn't
+    /// generate anything new since preexisting types are skipped in
+    /// `define_live_types`.
+    ///
+    /// This function will trim the sets on `self` to only retain those types
+    /// which exports refer to that come from imports.
+    fn remove_types_redefined_by_exports(&mut self, resolve: &Resolve, world: WorldId) {
```

Comment:
Alternatively, can we add a fourth parameter to `InterfaceGenerator` functions like `type_resource` named `direction: Direction`. Then we can generate types according to this direction, to avoid this specific logic of removing redefined types by exports?